File: mobile/lib/sesion_nativa.dart
/// Cómo llega el token a la página sin quedarse escrito en el WebView.
///
/// ## El problema, dicho con precisión
///
/// La web guarda su token en `localStorage` y tiene un motivo bueno para no
/// usar una cookie `HttpOnly`: el mismo token hay que meterlo en la URL del
/// WebSocket de colaboración, y una cookie `HttpOnly` no es legible desde el
/// código que construye esa URL. Está explicado en
/// `frontend/src/services/sesion.tsx` y no se toca.
///
/// Lo que cambia dentro de la app es dónde acaba ese `localStorage`: en un
/// fichero del directorio de datos, `app_webview/Default/Local Storage`. Un
/// token con semanas de vigencia guardado ahí sale del teléfono en cualquier
/// copia de seguridad del sistema y se lee entero con root. En un navegador de
/// escritorio no hay alternativa; en un teléfono sí, y no aprovecharla sería
/// dejar el punto más débil justo donde es más fácil de arreglar.
///
/// ## Cómo se resuelve
///
/// El token lo custodia la carcasa nativa, cifrado por el Keystore, y la página
/// lo recibe por este puente **cada vez que arranca, en memoria**. La web lo
/// pone en la cabecera y en la URL del socket igual que siempre; lo único que
/// no hace es escribirlo. Cuando el WebView se cierra, en su disco no queda
/// nada que robar.
///
/// El sentido contrario también importa y es la mitad que se olvida: la página
/// es la única que se entera de que el servidor ha dejado de aceptar la sesión
/// —le contesta 401 a una petición que sí llevaba token—, y si no lo dijera, el
/// almacén seguro conservaría para siempre una llave que ya no abre nada.
///
/// La otra mitad está en `frontend/src/services/sesion-nativa.ts`. Los dos
/// nombres de abajo son el contrato entre las mitades, y no hay compilador que
/// lo compruebe.
library;

import 'dart:convert';

import 'package:webview_flutter/webview_flutter.dart';

import 'sesion/token.dart';

/// Canal por el que la página pregunta y avisa (web → Flutter).
const String nombreCanalSesion = 'SesionNativa';

/// Función global que se llama para contestar (Flutter → web).
const String nombreReceptorSesion = '__sesionNativa';

/// El JavaScript que entrega una respuesta a la página.
///
/// Doblemente codificado, igual que los demás puentes: dentro viaja un token,
/// que es texto arbitrario firmado por el servidor, y un nombre de usuario, que
/// lo escribió una persona. Cualquiera de los dos puede traer una comilla, y sin
/// esto lo que se ejecutaría en la página sería un error de sintaxis —o, con un
/// nombre elegido a mala idea, algo peor.
String guionRespuestaSesion(Map<String, Object?> mensaje) {
  final carga = jsonEncode(jsonEncode(mensaje));
  return 'if (typeof window.$nombreReceptorSesion === "function") '
      'window.$nombreReceptorSesion($carga);';
}

/// El mensaje que se le manda a la página con la sesión que haya.
///
/// Está separado del puente para poder comprobar en una prueba qué se le cuenta
/// exactamente a la web, que es donde está el riesgo: mandar de más aquí es
/// devolver al `localStorage` lo que se acaba de sacar de él.
Map<String, Object?> mensajeDeSesion(SesionGuardada? sesion) {
  if (sesion == null) {
    // Sin sesión se contesta igual, con el token a `null`. Callar dejaría a la
    // página esperando para siempre, y una pantalla que no termina de cargar es
    // peor diagnóstico que una que dice que hay que entrar.
    return {'tipo': 'sesion', 'token': null};
  }
  return {
    'tipo': 'sesion',
    'token': sesion.token,
    'usuario': {
      'id': sesion.usuarioId,
      'email': sesion.email,
      'nombre': sesion.nombre,
    },
  };
}

/// Por qué se ha quedado la app sin sesión.
enum MotivoDeSalida {
  /// El servidor contestó 401: el token existe pero ya no lo acepta.
  caducada,

  /// Alguien pulsó «salir» dentro de la página.
  aPeticion,
}

/// Conecta la sesión de la página con la sesión guardada del teléfono.
///
/// Se registra junto a los demás puentes al crear el controlador, y **tiene que
/// estar antes de `loadRequest`**: la página pregunta por el token durante su
/// arranque, y un canal que aparezca después de que la página haya cargado es
/// un canal que la página ya dio por ausente. El síntoma sería la pantalla de
/// acceso de la web dentro de una app que acaba de pedir la contraseña.
class PuenteSesion {
  PuenteSesion({required this.alPerderLaSesion});

  /// Qué hace la carcasa cuando la sesión deja de valer.
  ///
  /// Lo decide `main.dart` y no este fichero: aquí no se sabe si hay una
  /// pantalla de acceso a la que volver, y un puente que navegue por su cuenta
  /// es un puente que no se puede probar.
  final void Function(MotivoDeSalida motivo) alPerderLaSesion;

  WebViewController? _web;
  SesionGuardada? _sesion;

  /// Por dónde salen las respuestas. Se desvía en las pruebas.
  void Function(Map<String, Object?>)? _responderPrueba;

  /// La sesión que se le entregará a la página cuando la pida.
  ///
  /// Se pone antes de registrar el puente. Puede volver a ponerse: al renovar
  /// la sesión sin recargar el WebView, la siguiente pregunta debe recibir la
  /// nueva y no la de antes.
  // ignore: use_setters_to_change_properties
  void usar(SesionGuardada? sesion) {
    _sesion = sesion;
  }

  void registrar(WebViewController web) {
    _web = web;
    web.addJavaScriptChannel(
      nombreCanalSesion,
      onMessageReceived: (mensaje) => recibir(mensaje.message),
    );
  }

  void responderA(void Function(Map<String, Object?>) destino) {
    _responderPrueba = destino;
  }

  void _responder(Map<String, Object?> mensaje) {
    final destino = _responderPrueba;
    if (destino != null) {
      destino(mensaje);
      return;
    }
    _web?.runJavaScript(guionRespuestaSesion(mensaje));
  }

  /// Procesa un mensaje del canal.
  ///
  /// Público para que las pruebas lo alimenten directamente: lo que hay que
  /// comprobar son las formas malas del mensaje y qué se contesta a cada una, y
  /// nada de eso necesita un teléfono.
  void recibir(String crudo) {
    // Viene de nuestra propia página y se valida igual. Una excepción dentro
    // del callback del canal no la recoge nadie, y el síntoma sería una app que
    // pide la contraseña y luego enseña otra pantalla de acceso, sin ningún
    // error por medio.
    Object? decodificado;
    try {
      decodificado = jsonDecode(crudo);
    } on FormatException {
      return;
    }
    if (decodificado is! Map<String, Object?>) return;

    switch (decodificado['tipo']) {
      case 'pedir':
        _responder(mensajeDeSesion(_sesion));
      case 'caducada':
        _sesion = null;
        alPerderLaSesion(MotivoDeSalida.caducada);
      case 'salir':
        _sesion = null;
        alPerderLaSesion(MotivoDeSalida.aPeticion);
      // Cualquier otra cosa se ignora en silencio: es lo que llegaría de un
      // frontend más nuevo que el APK, y contestar con un error convertiría una
      // incompatibilidad benigna en ruido en la consola.
    }
  }
}

File: frontend/src/services/sesion-nativa.ts
/**
 * De dónde sale el token cuando la página corre dentro de la app.
 *
 * ## Qué cambia respecto al navegador
 *
 * En un navegador la sesión se guarda en `localStorage`, con el motivo escrito
 * en `sesion.tsx`: el token hace falta también para construir la URL del
 * WebSocket, y una cookie `HttpOnly` no es legible desde ahí. Esa decisión es
 * correcta en un navegador y no se toca.
 *
 * Dentro del WebView de Android, sin embargo, `localStorage` es un fichero
 * dentro del directorio de datos de la aplicación. Un token con semanas de
 * vigencia guardado ahí viaja en cualquier copia de seguridad del sistema y se
 * lee entero en un teléfono con root. El teléfono tiene un sitio mejor —el
 * Keystore— y la carcasa nativa sabe usarlo, así que dentro de la app **el
 * dueño del token es Flutter** y la página lo recibe prestado para la vida de
 * la pestaña.
 *
 * De ahí la regla que gobierna este fichero, y que es lo único que hay que
 * recordar al leerlo: **si hay puente, la página no escribe el token en ningún
 * sitio.** Ni al entrar, ni al recibirlo, ni al renovarlo.
 *
 * ## La otra mitad
 *
 * `mobile/lib/sesion_nativa.dart`. Los dos nombres de abajo son el contrato
 * entre las mitades y no hay compilador que lo compruebe, así que están
 * afirmados en las pruebas de los dos lados.
 *
 * Este fichero no toca el DOM ni el estado de React: solo decide y traduce,
 * para poder probarlo sin navegador. El cableado está en `sesion.tsx`.
 */

/** Canal por el que la página pregunta y avisa (web → Flutter). */
export const CANAL_SESION = 'SesionNativa';

/** Función global que Flutter llama para contestar (Flutter → web). */
export const RECEPTOR_SESION = '__sesionNativa';

interface PuenteNativo {
  postMessage(mensaje: string): void;
}

interface VentanaConSesion {
  [CANAL_SESION]?: PuenteNativo;
  [RECEPTOR_SESION]?: (crudo: string) => void;
}

/**
 * La ventana, si la hay, vista como algo que puede tener el puente.
 *
 * Devuelve `null` durante el renderizado en servidor y en las pruebas, que es
 * lo que permite que el resto del módulo no tenga que preguntarse si existe
 * `window`.
 */
export function ventanaConSesion(): VentanaConSesion | null {
  return typeof window === 'undefined' ? null : (window as unknown as VentanaConSesion);
}

/**
 * ¿Manda la carcasa nativa sobre la sesión?
 *
 * Se comprueba por la presencia del canal y no por el `userAgent`. El
 * `userAgent` de un WebView se puede parecer al de Chrome tanto como se quiera
 * —y de hecho se parece—, mientras que el canal existe exactamente cuando hay
 * alguien nativo escuchando al otro lado. Preguntar por lo que se va a usar, y
 * no por quién se dice ser, es la diferencia entre esto y una heurística.
 */
export function haySesionNativa(): boolean {
  return Boolean(ventanaConSesion()?.[CANAL_SESION]);
}

/** El usuario tal y como lo manda el lado nativo. */
export interface UsuarioNativo {
  id: string;
  email: string;
  nombre: string;
}

/** Lo que llega por el receptor cuando se pide la sesión. */
export interface SesionNativa {
  /** El token, o `null` si la carcasa no tiene ninguna sesión guardada. */
  token: string | null;
  usuario: UsuarioNativo | null;
}

/**
 * Lee el mensaje que manda Flutter, o devuelve `null` si no es de los nuestros.
 *
 * No lanza en ningún caso, y la razón es la misma que en `respaldo.ts`: quien
 * llama a esto está dentro de la función global que invoca Flutter por
 * `runJavaScript`, y una excepción ahí no la recoge nadie —sube hasta el puente
 * y se pierde—. El síntoma de dejarla escapar sería una aplicación parada para
 * siempre en «comprobando la sesión», sin nada en la consola.
 *
 * Se distingue el mensaje mal formado (`null`) de la respuesta legítima sin
 * sesión (`token: null`), y la distinción no es cosmética: la segunda significa
 * «hay que entrar» y la primera no significa nada, así que tratarlas igual
 * mandaría a la pantalla de acceso por un mensaje que ni siquiera era para
 * nosotros.
 */
export function leerMensajeDeSesion(crudo: string): SesionNativa | null {
  let mensaje: unknown;
  try {
    mensaje = JSON.parse(crudo);
  } catch {
    return null;
  }
  if (typeof mensaje !== 'object' || mensaje === null) return null;
  const cuerpo = mensaje as Record<string, unknown>;
  if (cuerpo['tipo'] !== 'sesion') return null;

  const token = cuerpo['token'];
  if (token === null || token === undefined || token === '') {
    return { token: null, usuario: null };
  }
  if (typeof token !== 'string') return null;

  return { token, usuario: leerUsuario(cuerpo['usuario']) };
}

/**
 * El perfil que acompaña al token, con lo que falte puesto en blanco.
 *
 * Un token sin perfil legible sigue siendo un token bueno: la página puede
 * pedir `GET /api/auth/yo` y enterarse. Así que un perfil incompleto no invalida
 * la sesión, solo deja campos vacíos que se rellenarán al hablar con el
 * servidor. Rechazar la sesión entera por un nombre ausente sería pedir la
 * contraseña por no saber cómo llamar a quien ya está dentro.
 */
function leerUsuario(valor: unknown): UsuarioNativo | null {
  if (typeof valor !== 'object' || valor === null) return null;
  const u = valor as Record<string, unknown>;
  const id = u['id'];
  if (typeof id !== 'string' || id === '') return null;
  return {
    id,
    email: typeof u['email'] === 'string' ? u['email'] : '',
    nombre: typeof u['nombre'] === 'string' ? u['nombre'] : '',
  };
}

/* -------------------------------------------------------------------------- */
/* Lo que se le dice al lado nativo                                            */
/* -------------------------------------------------------------------------- */

export type AvisoNativo = 'pedir' | 'caducada' | 'salir';

/**
 * Manda un aviso por el canal. No hace nada fuera de la app.
 *
 * Que no haga nada es lo correcto y no una omisión: en un navegador de
 * escritorio la sesión la lleva `localStorage` y no hay nadie a quien avisar.
 */
export function avisarAlNativo(aviso: AvisoNativo): void {
  ventanaConSesion()?.[CANAL_SESION]?.postMessage(JSON.stringify({ tipo: aviso }));
}

/**
 * Cuánto se espera la respuesta del lado nativo antes de seguir sin ella.
 *
 * Hace falta un tope porque el canal no garantiza respuesta. `postMessage`
 * entrega y devuelve; si el lado de Flutter se cae, o el mensaje de vuelta pasa
 * del tope del Binder, o sencillamente el APK es más viejo que esta página y no
 * conoce el tipo «pedir», aquí no llega nada y **tampoco llega ningún error**.
 * Sin tope, el síntoma sería la aplicación detenida para siempre en la pantalla
 * de carga, que es el peor fallo posible: no se puede ni entrar ni leer lo que
 * ya estaba guardado en el teléfono.
 *
 * Cuatro segundos porque lo que hay al otro lado es una lectura del almacén
 * cifrado ya hecha —la carcasa tiene el token en memoria antes de abrir el
 * WebView— y eso son milisegundos. El margen es para el arranque en frío de un
 * teléfono lento, no para una operación lenta.
 */
export const ESPERA_SESION_NATIVA_MS = 4000;

/* -------------------------------------------------------------------------- */
/* La decisión que ordena el arranque                                          */
/* -------------------------------------------------------------------------- */

/** Lo que se sabe al arrancar, antes de enseñar nada. */
export interface SituacionArranque {
  /** ¿Existe el canal? */
  hayPuente: boolean;
  /** Lo que hubiera en `localStorage`, si es que se llegó a mirar. */
  tokenGuardado: string | null;
}

export type OrigenDelToken = 'nativo' | 'navegador' | 'ninguno';

/**
 * De dónde hay que sacar el token en este arranque.
 *
 * Hay una regla y tiene una consecuencia que conviene ver escrita: **con puente,
 * lo que hubiera en `localStorage` se ignora**, aunque haya algo.
 *
 * Podría parecer una reserva útil —«si el nativo no contesta, al menos
 * tenemos esto»— y es justo lo contrario. Ese resto solo puede venir de una
 * versión anterior del APK, la que guardaba el token en el WebView; darle uso
 * sería resucitar precisamente el almacenamiento del que se está huyendo, y
 * hacerlo en silencio, de modo que nadie notaría que la sesión volvió a estar
 * en claro en el disco. Por eso `sesion.tsx` no solo lo ignora: lo borra.
 */
export function origenDelToken({ hayPuente, tokenGuardado }: SituacionArranque): OrigenDelToken {
  if (hayPuente) return 'nativo';
  return tokenGuardado ? 'navegador' : 'ninguno';
}
